### DISCUSSION TITLE
renovate not updating package-lock properly when using NPM7

### DISCUSSION THREAD

**Discussion by rvillane** [renovatebot/renovate#8706] [Request Help]
**What Renovate type, platform and version are you using?**
gitHub.com
renovate CLI
Node v12.20.0
NPM v7.5.4
Renovate v24.49.7

**Describe the bug**
When used NPM 6, package-lock.json was updated along with package.json changes, however, when using NPM 7, only package.json is getting updated.

Example: https://github.com/rvillane/renovate-pinning/pull/25 

**Relevant debug logs**
{"name":"renovate","hostname":"lr90pmrvm-lp.xxxxxxxxxx","pid":20490,"level":20,"logContext":"0PssNFlm6","repository":"rvillane/renovate-pinning","branch":"renovate/all-minor-patch","command":"npm install --package-lock-only --ignore-scripts --no-audit","msg":"Executing command","time":"2021-02-15T21:54:22.111Z","v":0}
{"name":"renovate","hostname":"lr90pmrvm-lp.xxxxxxxx","pid":20490,"level":20,"logContext":"0PssNFlm6","repository":"rvillane/renovate-pinning","branch":"renovate/all-minor-patch","cmd":"npm install --package-lock-only --ignore-scripts --no-audit","durationMs":6371,"stdout":"\nup to date in 6s\n\n35 packages are looking for funding\n  run `npm fund` for details\n","stderr":"npm WARN old lockfile \nnpm WARN old lockfile The package-lock.json file was created with an old version of npm,\nnpm WARN old lockfile so supplemental metadata must be fetched from the registry.\nnpm WARN old lockfile \nnpm WARN old lockfile This is a one-time fix-up, please be patient...\nnpm WARN old lockfile \n","msg":"exec completed","time":"2021-02-15T21:54:28.482Z","v":0}
{"name":"renovate","hostname":"lr90pmrvm-lp.xxxxxxx","pid":20490,"level":20,"logContext":"0PssNFlm6","repository":"rvillane/renovate-pinning","branch":"renovate/all-minor-patch","msg":"package-lock.json hasn't changed","time":"2021-02-15T21:54:28.505Z","v":0}

**Have you created a minimal reproduction repository?**

Yes, https://github.com/rvillane/renovate-pinning, it contains [renovate-config.js](https://github.com/rvillane/renovate-pinning/blob/master/renovate-config.js) file, used to run the bot against this repo. 

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [X] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**
Seems to be an issue specific to the recently released NPM 7, package-lock.json on this repo has "lockfileVersion": 2.

If this is just me using wrong settings, please let me know.

**Comment by viceice** [comment]
To get renovate working with multiple node versions you need to use `binarySource=docker`, this will allow renovate to start a docker sidecar container with the required node and npm version.

https://docs.renovatebot.com/self-hosted-configuration/#binarysource

**Reply by rvillane** [reply]
should I use that setup even if I'm running renovate as a CLI standalone process ?  currently I don't use it as docker container

**Reply by viceice** [reply]
yes, otherwise you will get artifact issues all the time your global node and npm do not match the repo requested versions

**Reply by viceice** [reply]
you can use the rootless docker if you don't want to run docker daemon with root priviledges

**Reply by rvillane** [reply]
something I don't fully understand is that I have NPM 7 in the computer where renovate runs and the repo has NPM 7 package-lock, where is the mismatch in this case ?

**Reply by ext** [reply]
I currently use the docker image `renovate/renovate` and get `Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?` when I set `binarySource` to `docker`. I assume that is because the image is not `dind`. How can I get it working with the docker image?

I found this discussion when I was searching for issues with `package-lock.json` no longer being updated since NPM 7.

**Reply by viceice** [reply]
when using `binarySource=docker` you should use `renovate/renovate:slim` image and pass the docker socket as volume

The renovate doesn't run dind itself, but is able to use a dind sidecar container (not rootless). For this to work you need to mount the host `/tmp` to both container (you can use a volume too, but paths in containers need to match [baseDir](https://docs.renovatebot.com/self-hosted-configuration/#basedir))

PLease open a new discussion for further assistance.

**Comment by rarkins** [comment]
The key log line is `package-lock.json hasn't changed`.  Makes me wonder if `--package-lock-only` is broken  on npm 7 or something like that.

**Reply by rvillane** [reply]
I find this "WARN old lockfile The package-lock.json file was created with an old version of npm" message highly suspicious, because the actual package-lock.json included in the scanned repo was generated using NPM 7, however, the transient dependencies were generated using NPM 6 and I'm unclear if that combination causing issues when running npm install with --package-lock-only flag.

**Reply by rvillane** [reply]
I ran a quick experiment, upgrading @google-cloud/monitoring library from "1.7.0" to "2.2.0" in package.json and ran the following command:

`npm install --package-lock-only --ignore-scripts --no-audit
`
and it definitely did produce changes in package-lock.js file reflecting the new @google-cloud/monitoring dependencies.

I ran this experiment from the same machine where I ran the bot, then is using the same node and npm versions.

Any ideas ?

**Reply by viceice** [reply]
it must be something with your infra structure, see: 

- https://github.com/viceice-tests/renovate-pinning/pull/14
- https://github.com/viceice-tests/renovate-pinning/runs/1925508815?check_suite_focus=true

**Reply by rarkins** [reply]
What we know is that WARN is coming from npm, so it thinks it's seeing an npm6 lock file. But I don't know why that would be.

**Reply by viceice** [reply]
I think it was a temporary issue. maybe he used an older npm v7?

```
DEBUG: Executing command (repository=viceice-tests/renovate-pinning, branch=renovate/google-cloud-monitoring-2.x)
       "command": "npm install --package-lock-only --ignore-scripts --no-audit"
DEBUG: exec completed (repository=viceice-tests/renovate-pinning, branch=renovate/google-cloud-monitoring-2.x)
       "cmd": "npm install --package-lock-only --ignore-scripts --no-audit",
       "durationMs": 1155,
       "stdout": "\nup to date in 1s\n",
       "stderr": ""
DEBUG: package-lock.json needs updating (repository=viceice-tests/renovate-pinning, branch=renovate/google-cloud-monitoring-2.x)
DEBUG: Updated 1 lock files (repository=viceice-tests/renovate-pinning, branch=renovate/google-cloud-monitoring-2.x)
       "updatedArtifacts": ["package-lock.json"]
```

**Reply by rvillane** [reply]
I'm using npm v7.5.4.  @viceice , what npm version are you using ?

**Reply by viceice** [reply]
latest installable, see my fork, there it's working as expected, can you delete your branch and try again?

**Reply by rvillane** [reply]
I deleted the branches and PR's and tried again but got the same result (no package-lock update).  Could the renovate version make a difference, in my local environment I have renovate v24.49.7.  What version are you using ?

**Reply by viceice** [reply]
sure, https://github.com/viceice-tests/renovate-pinning/runs/1925550101?check_suite_focus=true

last run was on 24.53.0 with node 12.20.1 and npm 7.5.4

**Reply by rvillane** [reply]
yup, I see it worked for you. I deleted branches, repeated the execution using those exact same versions with no luck.  Definitely something on my local environment.  Will try with a local GHE repo instead of public GH repo, to verify if firewalls could be interfering here.

**Reply by geraintwhite** [reply]
I'm having the same thing in my self-hosted setup.

I run renovate in docker (`docker run renvovate/renovate`) and recently switched my project from npm 6 to npm 8. Upon doing so renovate would rewrite my package-lock.json to lockfileVersion 1.

I added `binarySource: 'docker'` to my renovate config and now I'm seeing the log `package-lock.json hasn't changed` and getting no package-lock.json changes in PRs.

**Reply by viceice** [reply]
you need to mount the docker socket or use dind. You should use the slim image when in docker mode.

**Reply by geraintwhite** [reply]
I was mounting the docker socket, but not /tmp (discovered [here](https://github.com/renovatebot/renovate/issues/3153#issuecomment-524881083)) which was causing the issue.
